### PR TITLE
Catch socket timeouts and clear dead connection (refs #18)

### PR DESCRIPTION
**Pull request by justinfx**
Re #18 , this MR catches a socket timeout in the request and clears the cached connection instance. This allows for the following request to establish a new connection instead of hitting the dead connection.

Not ready for merge yet. While manual reproduction mentioned in #18 now works correctly with this change, the tests are not properly failing when the patch is not applied. I am looking for some feedback to understand how to adjust the test to behave the same way a normal production usage would function with the `Http` class.

### CODE REVIEW

**Comment by temoto** [comment]
I'm also trying to get proper test without success so far.

**Comment by temoto** [comment]
@justinfx checking response content on second request shows more problems than wanted.

```python
def test_timeout_subsequent():
    print('')
    timeout = 0.5
    sync_delay = 0.1

    def handler(read):
        request = read()
        print('- request client={} raw={}'.format(request.client_addr, request.raw))
        time.sleep(timeout + sync_delay)
        response = tests.http_response_bytes(status=500)
        for c in response:
            time.sleep(0.001)
            yield c
        print('- request client={} raw={}'.format(request.client_addr, request.raw))
        # read()
        yield tests.http_response_bytes(status=200, body=b"correct")

    http = httplib2.Http(timeout=timeout)
    http.force_exception_to_status_code = True
    with tests.server_yield(handler, request_count=2) as uri:
        print('- test request 1, expect 408 timeout')
        response, _ = http.request(uri)
        print(response, _)
        assert response.status == 408
        # time.sleep(timeout + sync_delay + sync_delay)
        print('- test request 2, expect 200')
        response, content = http.request(uri)
        print(response, content)
        assert response.status == 200
        assert content == b"correct"
```

```shell
# sometimes it passed (incorrectly) sometimes not, this runs faster than full script/test
while ./venv-27/bin/pytest -svk timeout_sub tests/test_other.py ; do ; done
```

**Comment by temoto** [comment]
@justinfx checking second reponse body shows more errors than wanted.

```python
def test_timeout_subsequent():
    print('')
    timeout = 0.5
    sync_delay = 0.1

    def handler(read):
        request = read()
        print('- request raw={}'.format(request.raw))
        time.sleep(timeout + sync_delay)
        response = tests.http_response_bytes(status=500)
        for c in response:
            time.sleep(0.001)
            yield c
        request = read()
        print('- request raw={}'.format(request.raw))
        yield tests.http_response_bytes(status=200, body=b"correct")

    http = httplib2.Http(timeout=timeout)
    http.force_exception_to_status_code = True
    with tests.server_yield(handler, request_count=2) as uri:
        print('- test request 1, expect 408 timeout')
        response, _ = http.request(uri)
        print(response, _)
        assert response.status == 408
        # time.sleep(timeout + sync_delay + sync_delay)
        print('- test request 2, expect 200')
        response, content = http.request(uri)
        print(response, content)
        assert response.status == 200
        assert content == b"correct"
```

```
# run particular test is faster than full script/test
# loop because it failed/passed randomly
while ./venv-27/bin/pytest -svk timeout_sub tests/test_other.py ; do ; done
```

**Comment by temoto** [comment]
tcpdump says client closes connection after 0.5s without your patch

**Comment by temoto** [comment]
So we've been digging the wrong way, it actually works fine on Python2 and fails on Python3 (as was specified in #18 ), I incorrectly assumed problem should be observed on py2.

**Comment by temoto** [comment]
Bare sockets server gives predictable pass on py2 and predictable fail on py3. I tend to blame py2 random fails on "tech debt" lack of `socket.close` in test server.

So I guess final task is to get test server behave like one below. Leave it to me.

```
from __future__ import print_function
import socket
import time

slow_delay = 0.6


def fmt_response(status, body):
    return b'HTTP/1.1 {s}\r\ncontent-length: {cl}\r\n\r\n{b}'.format(
        s=status, cl=len(body), b=body)


def per_conn(c):
    while True:
        try:
            r = c.recv(8<<10)
        except socket.error as e:
            print('! recv err:', repr(e))
            c.close()
            return
        print('<', repr(r))
        if not r:
            c.close()
            return
        rs = per_request(r)
        c.sendall(rs)


def per_request(r):
    assert r.endswith(b'\r\n\r\n')
    words = r.split(b' ', 2)
    assert words[0] == b'GET'
    path = words[1].rsplit(b'/', 1)[1]
    print('path: ', repr(path))
    if path == 's':
        time.sleep(slow_delay)
        return fmt_response('200 OK', b'slow')
    elif path == 'f':
        return fmt_response('200 OK', b'fast')
    return fmt_response('404 Wrong Path', b'')


s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind(('127.0.0.1', 9091))
s.listen(10)
while True:
    c, _ = s.accept()
    per_conn(c)
```

**Comment by thehesiod** [comment]
@justinfx you only fixed it for py2, please also fix for py3 as waiting for the response can trigger a timeout and cause the same issue.

**Comment by justinfx** [comment]
I've added the fix to the python3 package and tests pass for python 3.6.6.
Surprisingly, the skipped `test_timeout_global` also passes for me on 3.6.6 so I am not sure why it is being skipped.

**Comment by temoto** [comment]
@httplib2/maintainers guys help. The problem and fix is "obviously correct". Test in patch does not work (no fail against code without fix), because test server is not surfacing the issue (my bad).

This is against best practices, but I think we should merge fix without test.

**Comment by justinfx** [comment]
@httplib2/maintainers bump

File: python2/httplib2/__init__.py [hunk 1]
```diff
@@ -2133,13 +2135,19 @@ def request(
                         cachekey,
                     )
         except Exception as e:
+            isTimeout = isinstance(e, socket.timeout)
```
Comment (temoto) [inline python2/httplib2/__init__.py]: Please change it to uniform snake case in both py2/3 and I'll merge it.

```suggestion
            is_timeout = isinstance(e, socket.timeout)
```

File: python2/httplib2/__init__.py [hunk 2]
```diff
@@ -2133,13 +2135,19 @@ def request(
...
+            if isTimeout:
```
Comment (temoto) [inline python2/httplib2/__init__.py]:
```suggestion
            if is_timeout:
```

File: python2/httplib2/__init__.py [hunk 3]
```diff
@@ -2133,13 +2135,19 @@ def request(
...
+                    conn.close()
+                    
             if self.force_exception_to_status_code:
                 if isinstance(e, HttpLib2ErrorWithResponse):
                     response = e.response
                     content = e.content
                     response.status = 500
                     response.reason = str(e)
-                elif isinstance(e, socket.timeout):
+                elif isTimeout:
```
Comment (temoto) [inline python2/httplib2/__init__.py]:
```suggestion
                elif is_timeout:
```

Surprised to say it, but Github is getting better each day after that MS deal.

**Comment by thehesiod** [comment]
will there be a release with this fix?

**Comment by temoto** [comment]
@thehesiod done, it's 0.12.1

**Pull request merged**